File: packages/guards/src/json-repair.ts
import type { Guard, GuardResult, GuardContext } from 'open-guardrail-core';

interface JsonRepairOptions {
  action: 'block' | 'warn' | 'fix';
}

function tryRepairJson(text: string): { repaired: string; wasFixed: boolean } | null {
  let trimmed = text.trim();

  // Already valid
  try {
    JSON.parse(trimmed);
    return { repaired: trimmed, wasFixed: false };
  } catch {
    // continue to repair
  }

  // Extract JSON from markdown code blocks
  const codeBlockMatch = trimmed.match(/```(?:json)?\s*\n?([\s\S]*?)\n?```/);
  if (codeBlockMatch) {
    trimmed = codeBlockMatch[1].trim();
    try {
      JSON.parse(trimmed);
      return { repaired: trimmed, wasFixed: true };
    } catch {
      // continue
    }
  }

  // Fix trailing commas before } or ]
  let fixed = trimmed.replace(/,\s*([}\]])/g, '$1');
  try {
    JSON.parse(fixed);
    return { repaired: fixed, wasFixed: true };
  } catch {
    // continue
  }

  // Fix missing closing brackets
  const opens = (trimmed.match(/[{[]/g) ?? []).length;
  const closes = (trimmed.match(/[}\]]/g) ?? []).length;
  if (opens > closes) {
    let attempt = trimmed;
    for (let i = 0; i < opens - closes; i++) {
      attempt += trimmed.lastIndexOf('[') > trimmed.lastIndexOf('{') ? ']' : '}';
    }
    try {
      JSON.parse(attempt);
      return { repaired: attempt, wasFixed: true };
    } catch {
      // continue
    }
  }

  // Fix single quotes to double quotes
  fixed = trimmed.replace(/'/g, '"');
  try {
    JSON.parse(fixed);
    return { repaired: fixed, wasFixed: true };
  } catch {
    // give up
  }

  return null;
}

export function jsonRepair(options: JsonRepairOptions): Guard {
  return {
    name: 'json-repair',
    version: '1.0.0',
    description: 'Validates and repairs malformed JSON output from LLMs',
    category: 'format',
    supportedStages: ['output'],
    async check(text: string, _ctx: GuardContext): Promise<GuardResult> {
      const start = performance.now();

      const result = tryRepairJson(text);

      if (!result) {
        return {
          guardName: 'json-repair',
          passed: false,
          action: options.action === 'fix' ? 'block' : options.action,
          latencyMs: Math.round(performance.now() - start),
          details: { reason: 'Could not parse or repair JSON' },
        };
      }

      if (!result.wasFixed) {
        return {
          guardName: 'json-repair',
          passed: true,
          action: 'allow',
          latencyMs: Math.round(performance.now() - start),
        };
      }

      // JSON was repaired
      if (options.action === 'fix') {
        return {
          guardName: 'json-repair',
          passed: true,
          action: 'override',
          overrideText: result.repaired,
          latencyMs: Math.round(performance.now() - start),
          details: { repaired: true },
        };
      }

      return {
        guardName: 'json-repair',
        passed: false,
        action: options.action,
        latencyMs: Math.round(performance.now() - start),
        details: { repaired: true, repairedText: result.repaired },
      };
    },
  };
}

File: python/open_guardrail/guards/json_repair.py
"""Try to parse JSON; attempt simple repairs."""

import json
import re
import time

from open_guardrail.core import GuardResult


def _try_repair(text: str) -> str | None:
    s = text.strip()
    # Fix trailing commas before } or ]
    s = re.sub(r",\s*([}\]])", r"\1", s)
    # Add missing closing braces/brackets
    opens = s.count("{") - s.count("}")
    s += "}" * max(opens, 0)
    brackets = s.count("[") - s.count("]")
    s += "]" * max(brackets, 0)
    try:
        json.loads(s)
        return s
    except (json.JSONDecodeError, TypeError):
        return None


class _JsonRepair:
    def __init__(
        self, *, action: str = "mask"
    ) -> None:
        self.name = "json-repair"
        self.action = action

    def check(
        self, text: str, stage: str = "output"
    ) -> GuardResult:
        start = time.perf_counter()

        # Valid JSON: pass through
        try:
            json.loads(text.strip())
            elapsed = (
                time.perf_counter() - start
            ) * 1000
            return GuardResult(
                guard_name="json-repair",
                passed=True,
                action="allow",
                latency_ms=round(elapsed, 2),
            )
        except (json.JSONDecodeError, TypeError):
            pass

        # Attempt repair
        repaired = _try_repair(text)
        elapsed = (time.perf_counter() - start) * 1000

        if repaired is not None:
            return GuardResult(
                guard_name="json-repair",
                passed=True,
                action="override",
                override_text=repaired,
                latency_ms=round(elapsed, 2),
                details={"repaired": True},
            )

        return GuardResult(
            guard_name="json-repair",
            passed=False,
            action="block",
            message="JSON is invalid and cannot"
            " be repaired",
            latency_ms=round(elapsed, 2),
        )


def json_repair(
    *, action: str = "mask"
) -> _JsonRepair:
    return _JsonRepair(action=action)
